libbitstream/bitstream.c

```diff
@@ -0,0 +1,313 @@
+#include <opae/log.h>
+#include <opae/properties.h>
+#include <opae/sysobject.h>
+
+#define METADATA_GUID     "58656F6E-4650-4741-B747-425376303031"
+#define METADATA_GUID_LEN 16
+#define GBS_AFU_IMAGE     "afu-image"
+#define BBS_INTERFACE_ID  "interface-uuid"
+
+/*
+ * Check for bitstream header and fill out bistream_info fields
+ */
+#define MAGIC       0x1d1f8680
+#define MAGIC_SIZE  4
+#define HEADER_SIZE 20
+STATIC int parse_metadata(opae_bitstream_info *info)
+{
+	unsigned i;
+
+	if (!info)
+		return -EINVAL;
+
+	if (info->data_len < HEADER_SIZE) {
+		OPAE_ERR("File too small to be GBS.");
+		return -1;
+	}
+
+	if (((uint32_t *)info->data)[0] != MAGIC) {
+		OPAE_ERR("No valid GBS header.");
+		return -1;
+	}
+
+	/* reverse byte order when reading GBS */
+	for (i = 0; i < sizeof(info->interface_id); i++)
+		info->interface_id[i] =
+			info->data[MAGIC_SIZE+sizeof(info->interface_id)-1-i];
+
+	info->rbf_data = &info->data[HEADER_SIZE];
+	info->rbf_len = info->data_len - HEADER_SIZE;
+
+	return 0;
+}
+
+STATIC fpga_result string_to_guid(const char *guid, fpga_guid *result)
+{
+	if (uuid_parse(guid, *result) < 0) {
+		OPAE_ERR("Error parsing guid: %s.", guid);
+		return FPGA_INVALID_PARAM;
+	}
+
+	return FPGA_OK;
+}
+
+STATIC fpga_result check_bitstream_guid(const uint8_t *bitstream)
+{
+	fpga_guid bitstream_guid;
+	fpga_guid expected_guid;
+	errno_t e;
+
+	e = memcpy_s(bitstream_guid, sizeof(bitstream_guid), bitstream,
+		     sizeof(fpga_guid));
+	if (EOK != e) {
+		OPAE_ERR("memcpy_s failed.");
+		return FPGA_EXCEPTION;
+	}
+
+	if (string_to_guid(METADATA_GUID, &expected_guid) != FPGA_OK)
+		return FPGA_INVALID_PARAM;
+
+	if (uuid_compare(bitstream_guid, expected_guid) != 0)
+		return FPGA_INVALID_PARAM;
+
+	return FPGA_OK;
+}
+
+STATIC fpga_result get_bitstream_ifc_id(const uint8_t *bitstream,
+					size_t bs_len,
+					fpga_guid *guid)
+{
+	fpga_result result = FPGA_EXCEPTION;
+	char *json_metadata = NULL;
+	uint32_t json_len = 0;
+	const uint8_t *json_metadata_ptr = NULL;
+	json_object *root = NULL;
+	json_object *afu_image = NULL;
+	json_object *interface_id = NULL;
+	errno_t e;
+
+	if (check_bitstream_guid(bitstream) != FPGA_OK)
+		goto out_free;
+
+	json_len = *((uint32_t *)(bitstream + METADATA_GUID_LEN));
+	if (json_len == 0) {
+		OPAE_MSG("Bitstream has no metadata.");
+		result = FPGA_OK;
+		goto out_free;
+	}
+
+	if (json_len > bs_len) {
+		OPAE_ERR("invalid bitstream metadata size.");
+		result = FPGA_EXCEPTION;
+		goto out_free;
+	}
+
+	json_metadata_ptr = bitstream + METADATA_GUID_LEN + sizeof(uint32_t);
+
+	json_metadata = (char *)malloc(json_len + 1);
+	if (json_metadata == NULL) {
+		OPAE_ERR("Could not allocate memory for metadata!");
+		return FPGA_NO_MEMORY;
+	}
+
+	e = memcpy_s(json_metadata, json_len + 1, json_metadata_ptr, json_len);
+	if (EOK != e) {
+		OPAE_ERR("memcpy_s failed.");
+		result = FPGA_EXCEPTION;
+		goto out_free;
+	}
+	json_metadata[json_len] = '\0';
+
+	root = json_tokener_parse(json_metadata);
```

Comment:
Here we're calling `json_tokener_parse` just to get the interface id. Would it make sense to parse this once and stash (in the opae_bitstream_info struct) either of these?
1. The resulting root json_object structure 
2. A metadata struct that encapsulates the metadata fields
Just a parsing optimization to think about.

Reply:
Will note as a TODO for @anandaravuri to evaluate.